core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/TransformCorrelatedScalarSubquery.java

```diff
@@ -124,7 +126,10 @@ public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Co
                     correlatedJoinNode.getInput(),
                     rewrittenSubquery,
                     correlatedJoinNode.getCorrelation(),
-                    producesSingleRow ? INNER : correlatedJoinNode.getType(),
+                    // EnforceSingleRowNode guarantees that exactly single matching row is produced
```

Comment:
if there is no EnforceSingleRowNode, we exit earlier:

```
        if (!searchFrom(subquery, context.getLookup())
                .where(EnforceSingleRowNode.class::isInstance)
                .recurseOnlyWhen(ProjectNode.class::isInstance)
                .matches()) {
            return Result.empty();
```

Reply:
I missed that. It wasn't immediately visible in the diff. However, that should not be necessary for this rule to work in general, as far as I can tell. We omit that check and rely on the cardinality property of the subplan. 

I guess my comment is that in order to future-proof this comment from non-local changes, we could just generalize it to say that the subplan is guaranteed to produce one row (regardless of how that's achieved), so this can be an inner join instead of a left join.

Reply:
I guess the rule insists on finding EnforceSingleRowNode because it also wants (needs?) to remove it.